oak_functions/loader/src/lookup_data.rs

```diff
@@ -0,0 +1,260 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::logger::Logger;
+use anyhow::{anyhow, Context};
+use hyper::{body::Bytes, client::connect::Connect, Body, Client, Request};
+use hyper_rustls::HttpsConnector;
+use log::Level;
+use prost::Message;
+use serde_derive::Deserialize;
+use std::{collections::HashMap, sync::RwLock, time::Instant};
+
+#[derive(Copy, Clone, Deserialize, Debug)]
+pub enum LookupDataAuth {
+    None,
+    GcpMetadataToken,
+}
+
+impl LookupDataAuth {
+    pub fn default() -> Self {
+        LookupDataAuth::None
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum LookupDataSource {
+    Http { url: String, auth: LookupDataAuth },
+    File(std::path::PathBuf),
+}
+
+/// An in-memory lookup store instance that can refresh its internal entries from the provided data
+/// file URL.
+///
+/// Entries in the data file path must be consecutive binary encoded and length delimited
+/// protobuf messages according to the definition in `/oak_functions/proto/lookup_data.proto`.
+pub struct LookupData {
+    lookup_data_source: Option<LookupDataSource>,
+    entries: RwLock<HashMap<Vec<u8>, Vec<u8>>>,
+    logger: Logger,
+}
+
+impl LookupData {
```

Comment:
Isn't this a bit backwards now? Or at least overlapping between the factory and this data. The initial idea of this data struct was that it would refresh itself from upstream; but now, I think it is the factory that should refresh its internal state of the data (perhaps behind an Arc), and provide that to the instances it creates. I think currently there are too many layers :)

Reply:
I think it is still the case that the lookup data will refresh itself. 

However, we will have to think that through, as the lookup data should stay consistent throughout a lookup---especially, when we start thinking about iterators, cf.  #2050.